### PR TITLE
[Good First Issue][TF FE]: Support complex tensors for Pack operation

### PR DESCRIPTION
**Pull request by hibahassan1**
### Details:
 - ***Addresses the issue** : [[Good First Issue][TF FE]: Support complex tensors for Pack operation #22954](https://github.com/openvinotoolkit/openvino/issues/22954)*
 - *Added support for complex tensors for pack operation*
 - *Let me know if any changes are required*

### CODE REVIEW

File: src/frontends/tensorflow_common/src/op/tempCodeRunnerFile.cpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    }
+
+    auto pack = make_shared<v0::Concat>(concat_inputs, axis);
+    set_node_name(node.get_name(), pack);
+    return {pack};
+}
+}  // namespace op
+}  // namespace tensorflow
+}  // namespace frontend
+}  // namespace ov
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/tempCodeRunnerFile.cpp:38]: why do we need it?

File: tests/layer_tests/tensorflow_tests/test_tf_Pack.py [hunk 2]
```diff
@@ -44,10 +40,50 @@ def create_pack_net(self, input_shape, input_num, axis, input_type):
...
+            tf_net = sess.graph_def
+        return tf_net, None
+
+    test_data_basic = [
+        dict(input_shape=[2, 4], input_num=2, axis=None),
+        dict(input_shape=[3, 1, 2], input_num=3, axis=1),
+    ]
+
+    @pytest.mark.parametrize("params", test_data_basic)
+    @pytest.mark.precommit_tf_fe
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Pack.py:83]:
```suggestion
    @pytest.mark.precommit
```

File: tests/layer_tests/tensorflow_tests/test_tf_Pack.py [hunk 3]
```diff
@@ -44,10 +40,50 @@ def create_pack_net(self, input_shape, input_num, axis, input_type):
     ]
 
     @pytest.mark.parametrize("params", test_data_basic)
-    @pytest.mark.precommit
+    @pytest.mark.precommit_tf_fe
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Pack.py:43]: revert unrelated changes to test with complex type above

File: src/frontends/tensorflow_common/src/op/tempCodeRunnerFile.cpp (hunk 1 above)
Comment (hibahassan1) [reply src/frontends/tensorflow_common/src/op/tempCodeRunnerFile.cpp:38]: This file is generated by the Code Runner extension in VS Code when running code snippets. It's a temporary file used to execute the code and is usually cleaned up automatically. It appears this file was unintentionally included in the commit. I'll remove this file. Sorry for any confusion caused.

**Comment by hibahassan1** [comment]
I have made the suggested changes

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 4]
```diff
@@ -16,22 +17,56 @@ namespace tensorflow {
...
+            auto complex_type_mark = as_type_ptr<ComplexTypeMark>(input.get_node_shared_ptr());
+            if (complex_type_mark) {
+                auto real_part = make_shared<v0::Unsqueeze>(complex_type_mark->input_value(0), axis_const);
+                extended_concat_inputs.push_back(real_part);
+                auto imag_part = make_shared<v0::Unsqueeze>(complex_type_mark->input_value(1), axis_const);
+                extended_concat_inputs.push_back(imag_part);
+            } else {
+                extended_concat_inputs.push_back(input);
+            }
+        }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:58]: is it possible to concatenate complex tensors without extraction of real and imaginary parts?
are you sure that type of operands can be different: sort of mix of real and complex types?

File: tests/layer_tests/tensorflow_tests/test_tf_Pack.py [hunk 5]
```diff
@@ -33,9 +31,7 @@ def create_pack_net(self, input_shape, input_num, axis, input_type):
             else:
                 tf.raw_ops.Pack(values=inputs)
             tf.compat.v1.global_variables_initializer()
-
             tf_net = sess.graph_def
-
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Pack.py:38]: please revert changes with deleted blank lines above. They are needed for readability.

File: src/frontends/tensorflow_common/src/op/pack.cpp (hunk 4 above)
Comment (hibahassan1) [reply src/frontends/tensorflow_common/src/op/pack.cpp:58]: I extracted the real and imaginary parts because there is no support for complex types. Could you please confirm if there are alternative methods I should explore to handle complex tensors more effectively within the framework? Regarding the second question, I am considering adding a separate implementation path for mixed-type tensors, treating real inputs as having zero imaginary parts and for tensors with only real types, adding an imaginary part with zeros to maintain consistency. Would this approach align with the project's requirements?

File: tests/layer_tests/tensorflow_tests/test_tf_Pack.py (hunk 5 above)
Comment (hibahassan1) [reply tests/layer_tests/tensorflow_tests/test_tf_Pack.py:38]: I will do that. Thank you

File: src/frontends/tensorflow_common/src/op/pack.cpp (hunk 4 above)
Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/pack.cpp:58]: Any idea how to support it without extraction real and imaginary parts?
Did you check TensorFlow documentation and what it says about support of mixed types?

Comment (hibahassan1) [reply src/frontends/tensorflow_common/src/op/pack.cpp:58]: In TensorFlow, when a real tensor and complex type tensor are stacked (as the newer version of Tensorflow changed the name of the pack operation to stack), an InvalidArgumentError is thrown. So the workaround was to convert the real tensor into a complex one to stack it together. A mixed tensor is often considered a complex type for ease as a TypeError will be thrown otherwise. I will refer to the Openvino operations which support complex types to see how it is handled there. Thank you.

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 6]
```diff
@@ -16,22 +17,38 @@ namespace tensorflow {
...
+
     for (int ind = 0; ind < num_size; ++ind) {
         auto in = node.get_input(ind);
+        auto complex_type_mark = as_type_ptr<ComplexTypeMark>(in.get_node_shared_ptr());
+        if (complex_type_mark) {
+            has_complex_input = true;
+            complex_part_type = complex_type_mark->get_complex_part_type();
+            in = complex_type_mark->input_value(0);
+        }
         concat_inputs.push_back(make_shared<v0::Unsqueeze>(in, axis_const));
```
[the remark this answers is not in the document]
Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/pack.cpp:38]: Please add test case for negative axis below

[the remark this answers is not in the document]
Comment (hibahassan1) [reply src/frontends/tensorflow_common/src/op/pack.cpp:45]: The new code handles negative axes by converting them to positive equivalents using the input tensor's rank. This ensures the axis is always valid

File: tests/layer_tests/tensorflow_tests/test_tf_Pack.py [hunk 7]
```diff
@@ -1,6 +1,3 @@
-# Copyright (C) 2018-2024 Intel Corporation
-# SPDX-License-Identifier: Apache-2.0
-
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Pack.py:3]: please recover copyright

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 8]
```diff
@@ -16,22 +18,44 @@ namespace tensorflow {
 namespace op {
 
 OutputVector translate_pack_op(const NodeContext& node) {
-    default_op_checks(node, 1, {"Pack", "PACK"});
+    default_op_checks(node, 1, {"Pack", "PACK"}, true);
     auto num_size = static_cast<int>(node.get_input_size());
 
     auto axis = node.get_attribute<int64_t>("axis", 0);
+    auto input_shape = make_shared<v0::ShapeOf>(node.get_input(0));
+    auto rank = static_pointer_cast<v0::Constant>(input_shape->output(0).get_node_shared_ptr())->cast_vector<int64_t>().size();
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:26]: In general case, this rank cannot be computed  during model conversion. It can be computed only during run-time (not in this code).
Any other trick, how to adjust `axis` without `ShapeOf`. What value shoud be subtracted for correctness:
axis -= some_value ?
Please find some value for negative case

Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/pack.cpp:26]: @hibahassan1, any update?

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 9]
```diff
@@ -16,22 +18,44 @@ namespace tensorflow {
...
+    if (axis < 0) {
+        axis += rank;
+    }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:29]:
```suggestion
>cast_vector<int64_t>().size();
    if (axis < 0) {
        axis -= 1;
    }
```

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 10]
```diff
@@ -16,22 +18,43 @@ namespace tensorflow {
 namespace op {
 
 OutputVector translate_pack_op(const NodeContext& node) {
-    default_op_checks(node, 1, {"Pack", "PACK"});
+    default_op_checks(node, 1, {"Pack", "PACK"}, true);
     auto num_size = static_cast<int>(node.get_input_size());
 
     auto axis = node.get_attribute<int64_t>("axis", 0);
+>cast_vector<int64_t>().size();
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:25]:
```suggestion
```

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 11]
```diff
@@ -16,22 +18,42 @@ namespace tensorflow {
...
     set_node_name(node.get_name(), pack);
+
+    if (has_complex_input) {
+        auto complex_pack = make_shared<ComplexTypeMark>(pack, complex_part_type);
+        return {complex_pack->output(0)};
+    }
+
     return {pack};
 }
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:56]:
```suggestion
```

**Review by rkazants — APPROVED** [review]

**Comment by rkazants** [comment]
build_jenkins

**Comment by rkazants** [comment]
build_jenkins

File: src/frontends/tensorflow_common/src/op/pack.cpp [hunk 12]
```diff
@@ -16,20 +18,39 @@ namespace tensorflow {
 namespace op {
 
 OutputVector translate_pack_op(const NodeContext& node) {
-    default_op_checks(node, 1, {"Pack", "PACK"});
+    default_op_checks(node, 1, {"Pack", "PACK"}, true);
     auto num_size = static_cast<int>(node.get_input_size());
 
     auto axis = node.get_attribute<int64_t>("axis", 0);
+    if (axis < 0) {
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/pack.cpp:25]:
```suggestion
    if (axis < 0 && is_type_ptr<ComplexTypeMark>(node.get_input(0).get_node_shared_ptr())) {
```

**Comment by rkazants** [comment]
build_jenkins

**Comment by hibahassan1** [comment]
Hello @rkazants ,

Sorry for the delay in my response. I got busy and didn't manage to address the issue in time. I also struggled with the solution for adjusting the axis without ShapeOf.

Thank you for your patience and for handling the updates. I appreciate it and am sorry for any inconvenience caused.
